spec.html

```diff
@@ -8794,6 +8834,24 @@ <h1>CreateBuiltinFunction ( _steps_, _internalSlotsList_ [ , _realm_ [ , _protot
       </emu-alg>
       <p>Each built-in function defined in this specification is created by calling the CreateBuiltinFunction abstract operation.</p>
     </emu-clause>
+
+    <emu-clause id="sec-createbuiltingeneratorfunction" aoid="CreateBuiltinGeneratorFunction">
+      <h1>CreateBuiltinGeneratorFunction ( _steps_ [ , _iteratorKind_ [ , _generatorPrototype_ [ , _generatorInternalSlotsList_ ]]]) </h1>
+      <p>The abstract operation CreateBuiltinGeneratorFunction create a <dfn id="sec-builtin-generator-function">built-in generator function</dfn> object. The generator object it returns can have extra internal slots.</p>
+      <p>The abstract operation CreateBuiltinGeneratorFunction takes arguments _steps_, and optional arguments _generatorInternalSlotsList_ (a List of internal slot names that the returned generator has), _generatorPrototype_ (the prototype of the returned generator, should inherit from %IteratorPrototype%). It performs the following steps when called:</p>
+      <emu-alg>
+        1. Assert: _steps_ is either a set of algorithm steps or other definition of a function's behaviour provided in this specification. It can contain Yield abstract operations to yield an IteratorResult object.
+        1. If _generatorPrototype_ is not present, set _generatorPrototype_ to %Iterator.prototype%.
+        1. Append &laquo; [[GeneratorState]], [[GeneratorContext]], [[IteratorKind]] &raquo; to _generatorInternalSlotsList_.
+        1. Let _wrappedSteps_ be:
+          1. Let _generator_ be ? OrdinaryObjectCreate(_generatorPrototype_, _generatorInternalSlotsList_).
+          1. Set _generator_.[[IteratorKind]] to _iteratorKind_.
+          1. Perform ? GeneratorStart(_generator_, _steps_).
+          1. Return _generator_.
+        1. Return ! CreateBuiltinFunction( _wrappedSteps_ ).
+      </emu-alg>
+      <p>Each built-in generator function defined in this specification is created by calling the CreateBuiltinGeneratorFunction abstract operation.</p>
```

Comment:
```
<emu-clause id="sec-createbuiltingeneratorfunctionsteps" aoid="CreateBuiltinGeneratorFunctionSteps">
      <h1>CreateBuiltinGeneratorFunctionSteps ( _steps_ [ , _iteratorBrand_ [ , _generatorInternalSlotsList_ ] ] )</h1>
      <p>The abstract operation CreateBuiltinGeneratorFunctionSteps takes arguments _steps_, and optional arguments _iteratorBrand_, _generatorPrototype_ (an Object), and _generatorInternalSlotsList_ (a List of internal slot names). It returns a new set of steps to used to create a <dfn id="sec-builtin-generator-function">built-in generator function</dfn> object by passing the new steps to the CreateBuiltinFunction. It returns a new set of steps described following:</p>
      <emu-alg>
        1. Assert: _steps_ is either a set of algorithm steps or other definition of a function's behaviour provided in this specification. It can contain uses of the Yield shorthand to yield an IteratorResult object.
        1. If _generatorPrototype_ is not present, set _generatorPrototype_ to %Iterator.prototype%.
        1. Append [[GeneratorState]], [[GeneratorContext]], [[IteratorBrand]] to _generatorInternalSlotsList_.
        1. Let _generator_ be ? OrdinaryObjectCreate(_generatorPrototype_, _generatorInternalSlotsList_).
        1. If _iteratorBrand_ presents, set _generator_.[[IteratorBrand]] to _iteratorBrand_.
        1. Perform ? GeneratorStart(_generator_, _steps_).
        1. Return _generator_.
      </emu-alg>
    </emu-clause>
```

what about this? And then let every built-in generator function wrap it's steps then call the CreateBuiltinFunction

Reply:
> what about this?

Hm. Something like that might be useful, but I'd want to see how it's used.

> And then let every built-in generator function wrap it's steps

Well, it doesn't actually wrap steps, it just hands them off to `GeneratorStart`.

Also, a built-in generator function (BIGF) wouldn't pass *its* steps to this operation, rather it would pass the associated iteration steps.

> then call the CreateBuiltinFunction

No, BIGFs don't call CreateBuiltinFunction. BIGFs are (presumably) created *by* calls to CreateBuiltinFunction.

Reply:
hmm maybe my word is not clear. generally my idea is

```javascript
SomeBIGF = CreateBuiltinFunction(
	CreateBuiltinGeneratorFunctionSteps( the steps X described below )
)
// Steps X:
1. do sth
2. do sth
3. ? Yield(_y_)
```

It's following your suggestion, does not create a new built-in function type but "prepare" the steps before CreateBuiltinFunction (add Generator related slots) to make it become a generator.

Reply:
Yup, I got the general idea. So now the specifics...

Even if `SomeBIGF = CreateBuiltinFunction(...)` is conceptually what happens, it's unlikely that that line (or anything like it) will appear in the spec, because intrinsic functions aren't created by explicit invocations of `CreateBuiltinFunction`, they're all created by a general statement in `CreateIntrinsics`. That statement tells the reader how to figure out what all the explicit invocations *would* be, based on the information that the spec gives elsewhere about each intrinsic function.

So in terms of what you've defined, the `%StringIterator%` clause would be something like:
```
%StringIterator% ( _s_ )
%StringIterator% is a built-in generator function.
It takes argument _s_ (a String) and returns a generator instance.
It performs the following steps when called:
  1. Let _steps_ be the iteration steps defined below.
  2. Return  CreateBuiltinGeneratorFunctionSteps(_steps_, ~String Iterator~, %StringIteratorPrototype%).

The iteration steps are:
  1. ....
  2. Repeat...
     a. Yield(...)
```

Two things:
- "CreateBuiltinGeneratorFunctionSteps" isn't a good name, because the operation doesn't create steps, let alone the steps of a BIGF. It creates a generator instance.
- The iteration steps need to reference the `_s_` passed to `%StringIterator%`, but they're a separate algorithm. The best way to handle this is probably with an abstract closure.

So then the steps of `%StringIterator%` would be something like:
```
  1. Let _closure_ be a new Abstract Closure with no parameters that captures _s_.
     It performs the following steps when called:
    a. ...
    b. Repeat...
      i. Yield(...)
  2. Return  CreateGeneratorInstance(_closure_, ~String Iterator~, %StringIteratorPrototype%).
```

Reply:
I'd like to abstract the declaration of generator functions so I don't need to write this every time but specify it once (like in CreateIntrinsics or CreateBuiltinGeneratorFunction) and use it anywhere in a simpler form.

```
  1. Let _closure_ be a new Abstract Closure with no parameters that capture _s_.
     It performs the following steps when called:
    a. ...
    b. Repeat...
      i. Yield(...)
  2. Return  CreateGeneratorInstance(_closure_, ~String Iterator~, %StringIteratorPrototype%).
```

The simpler form:

```
%X% is a built-in generator function with arguments _a_ and _b_.
It's returning generator has internal slots [[ X, Y, Z ]] and [[prototype]] be %XIteratorPrototype%.

The iteration steps: (implicitly create an Abstract Closure to capture all arguments of %X%)
1. ...
2. Yield ...
```

Does it ok?

Reply:
And if possible, I want to let the final form satisfy the form of ![image](https://camo.githubusercontent.com/84b38f72b9bc134b11c65c070bbd0cf58480bea8/68747470733a2f2f67632e67792f35383230353433332e706e67) so the iterator helpers proposal (cc @devsnek) will also feel ok to use it.

Reply:
> I'd like to abstract the declaration of generator functions so I don't need to write this every time but specify it once [...]

My advice is, specify it using existing spec constructs, and *then* make the case for introducing a new construct.